### PR TITLE
Test-time vol feature normalization (AdaBN) to close val→test vol_p gap

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Every experiment in this research program has shown a structural ~+7.6pp val→test gap in `volume_pressure_rel_l2_pct`. The model achieves val_vol_p ≈ 6.2–6.8% but test_vol_p ≈ 10.7–12.2%. This is confirmed to be a DATA DISTRIBUTION problem: the 50 test geometries are drawn from a different region of geometry space than the 400 train/val geometries.

**Hypothesis**: Adaptive Batch Normalization (AdaBN / Test-Time Normalization) applied to the volume branch can partially bridge this distribution gap by re-estimating volume feature statistics using the test geometry's own points at inference time — without any test labels. This is a classical domain adaptation technique that requires zero labeled test data and only ~O(1) forward passes per geometry.

**Mechanism**: During forward pass for test geometries, replace the running mean/variance in any BatchNorm or LayerNorm layers in the volume processing branch with statistics computed from the current geometry's volume point features. This effectively "re-normalizes" volume features to each test geometry's local distribution, reducing the covariate shift between train and test populations.

**Reference**: Wang et al. "Tent: Fully Test-Time Adaptation by Entropy Minimization" (ICLR 2021); Li et al. "Revisiting Batch Normalization For Practical Domain Adaptation" (ICLR 2017 Workshop).

## Instructions

This experiment adds **test-time volume feature normalization** to the existing model. The implementation lives entirely in the model forward pass and inference code.

### 1. Add `--use-test-time-vol-norm` CLI flag

In `target/train.py`, add a new argument:

```python
parser.add_argument("--use-test-time-vol-norm", action="store_true", default=False,
                    help="Apply test-time adaptive normalization to volume branch features (AdaBN)")
```

### 2. Implement Test-Time Volume Normalization

In the model's volume processing forward pass (wherever volume point features are aggregated — likely in the AB-UPT encoder's volume token pathway), add a test-time normalization step:

In the model file (likely `target/model.py` or equivalent), locate the volume hidden state computation. After computing the per-geometry volume hidden states `vol_h` (shape: `[B, N_vol, D]`), add:

```python
if self.use_test_time_vol_norm and not self.training:
    # Per-geometry, per-channel normalization using the geometry's own statistics
    # vol_h: [B, N_vol, D]
    mean = vol_h.mean(dim=1, keepdim=True)   # [B, 1, D]
    std  = vol_h.std(dim=1, keepdim=True) + 1e-6  # [B, 1, D]
    vol_h = (vol_h - mean) / std
    # Optionally apply learned scale/shift (initialized to identity)
    if hasattr(self, 'vol_norm_scale'):
        vol_h = vol_h * self.vol_norm_scale + self.vol_norm_bias
```

Add the learnable affine parameters to the model `__init__`:
```python
if use_test_time_vol_norm:
    self.vol_norm_scale = nn.Parameter(torch.ones(hidden_dim))
    self.vol_norm_bias  = nn.Parameter(torch.zeros(hidden_dim))
```

The key insight: during **training**, this normalization is OFF (the model learns without it). During **evaluation/test**, it's ON. This means the model learns representations that work with the train distribution, but at test time we adapt volume features to each test geometry's local distribution.

### 3. Wire flag through model construction

Pass `use_test_time_vol_norm=args.use_test_time_vol_norm` to the model constructor. Set `self.use_test_time_vol_norm = use_test_time_vol_norm` in `__init__`.

### 4. Run command

```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --epochs 35 \
  --lr 1e-4 \
  --weight-decay 0.005 \
  --optimizer lion \
  --grad-clip-norm 1.0 \
  --train-volume-points 16384 \
  --eval-volume-points 65536 \
  --use-test-time-vol-norm \
  --wandb-group nezuko-test-time-vol-norm \
  --kill-thresholds '{"ep3_val_abupt": 8.0, "ep3_val_vol_p": 5.0, "ep5_val_abupt": 7.5, "ep10_val_abupt": 7.2, "ep15_val_abupt": 6.80, "ep20_val_abupt": 6.70, "ep25_val_abupt": 6.65, "ep30_val_abupt": 6.60}'
```

### 5. Fallback if above approach doesn't apply cleanly

If the volume hidden states are not directly accessible before output projection (i.e., the model architecture doesn't have a clean `vol_h` tensor to intercept), implement a simpler variant: **per-geometry output rescaling**. After computing the raw `vol_p` predictions for a batch geometry, standardize the output predictions to zero mean unit variance using the geometry's own point distribution, then apply a learned global scale/shift:

```python
if self.use_test_time_vol_norm and not self.training:
    # vol_pred: [B, N_vol, 1]
    pred_mean = vol_pred.mean(dim=1, keepdim=True)
    pred_std  = vol_pred.std(dim=1, keepdim=True) + 1e-6
    vol_pred  = (vol_pred - pred_mean) / pred_std
```

Note: this output-level normalization changes the prediction scale, so the loss function during training must NOT have this enabled (keep `not self.training` guard). The ground-truth pressure fields have different statistics per geometry, so this effectively removes the per-geometry mean/scale before computing loss — equivalent to scale-invariant pressure prediction.

### 6. Kill gates

Apply standard kill gates:
- EP3: val_abupt ≤ 8.0%, val_vol_p ≤ 5.0%
- EP5: val_abupt ≤ 7.5%
- EP10: val_abupt ≤ 7.2%
- EP15: val_abupt ≤ 6.80%
- EP20: val_abupt ≤ 6.70%
- EP25: val_abupt ≤ 6.65%
- EP30: val_abupt ≤ 6.60%

## Baseline

Current best metrics (Wave SOTA — PR #740, W&B run `5x8wofzm`):
| Metric | val | test |
|--------|-----|------|
| abupt | ~6.46% | 7.5195% |
| vol_p | ~6.29% | **10.758%** |
| surf_p | ~3.30% | 3.881% |
| wall | ~4.84% | 7.061% |

Public AB-UPT targets: vol_p=6.08%, surf_p=3.82%, wall=7.29%

The structural val→test vol_p gap is +7.6pp. This is the primary target.

W&B run: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/5x8wofzm

Reproduce baseline:
```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --epochs 35 --lr 1e-4 --weight-decay 0.005 --optimizer lion \
  --grad-clip-norm 1.0 --train-volume-points 16384 --eval-volume-points 65536
```

## Expected outcome

If successful: test_vol_p drops below 10.758% (ideally toward 9–10%) while val metrics remain stable. The val→test gap should narrow from ~+7.6pp.

If the feature-level normalization doesn't help: confirms that the distributional gap is not addressable via inference-time feature re-centering, which would point toward geometry embedding / conditioning approaches instead.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: No acknowledgment on PR #1044 (AdaBN / Test-Time Normalization) yet.

Please acknowledge this assignment and:
1. Confirm you are starting implementation of the AdaBN inference pass
2. Post W&B run ID once launched

**EP3 kill gate**: val_abupt ≤8.0% AND val_vol_p ≤5.0%.

The implementation is a test-time inference modification — at eval, for each test geometry: freeze backbone params, run 10 forward passes with BN in train mode (batch of augmented point subsets), then switch to eval mode for the final prediction. Post EP1 val metrics as they land.

**Comment by morganmcg1** [comment]
ADVISOR: SECOND ESCALATION — No response received to my previous comment. This PR has been sitting with zero student activity since it was assigned. The AdaBN (Adaptive Batch Normalization) experiment (PR #1044) needs to start immediately.

To recap the assignment:
- Implement test-time vol feature normalization: during eval, compute per-geometry mean/std over the vol token sequence and normalize, then apply learnable `vol_norm_scale` / `vol_norm_bias` params (initialized to 1/0)
- Flag: `--use-test-time-vol-norm`
- W&B group: `nezuko-test-time-vol-norm`
- Kill gate EP3: val_abupt ≤ 8.0% AND val_vol_p ≤ 5.0%

Note: You have two PRs active (#1044 and #1043 Perceiver-IO). Prioritize getting BOTH running as soon as possible. Acknowledge this comment and post the W&B run ID once launched. This is urgent.

**Comment by morganmcg1** [comment]
ADVISOR: THIRD ESCALATION — PR #1044 (AdaBN / Test-Time Normalization, nezuko) still has zero student activity.

You have two active PR assignments (#1043 and #1044). Both must be running simultaneously. These are not optional.

Post an acknowledgment and W&B run ID for BOTH PRs within 15 minutes. If you need clarification on the implementation spec, ask now — do not stay silent.

The implementation for #1044 is straightforward: add a test-time vol feature normalization pass using per-geometry mean/std of the vol token sequence, with learnable scale/bias params initialized to identity. See the full spec in the PR body.

**Comment by morganmcg1** [comment]
ADVISOR: FOURTH AND FINAL ESCALATION — This is the last notice before this PR is reassigned.

PR #1044 (AdaBN / Test-Time Vol Feature Normalization) has had ZERO student activity since it was assigned. Three prior escalations have been ignored. You have 30 minutes from this comment to post an acknowledgment AND a W&B run ID showing the experiment is running.

If no response by then, this PR will be closed and the experiment reassigned to another student.

The implementation is a test-time-only modification — no changes to training loop, just an inference pass. It is one of the simplest experiments in the current wave. There is no valid reason for zero progress.

If you are blocked on something, ask NOW. That is what I am here for.

**Comment by morganmcg1** [comment]
ADVISOR: FOURTH AND FINAL ESCALATION — PR #1044 (AdaBN / Test-Time Normalization, nezuko)

This is the fourth time I have asked for acknowledgment. There has been zero student activity on this PR since it was assigned.

**Immediate action required:**
- If you cannot start this PR, comment NOW explaining why (GPU conflict, implementation blocker, etc.)
- If you ARE starting this PR, post the W&B run ID immediately

**Context:** You currently have PR #1043 (SlotVolAttn tay run `wbjumwbk`) running. PR #1044 is your drivaerml-long assignment. Both should be running simultaneously on separate GPU allocations.

**If I do not receive a response within the next 2 hours, I will close this PR and reassign the experiment to another student.** The hypothesis is still worth testing and will not be abandoned — it will simply be assigned elsewhere.

Note: This is not about blame. If there is a technical or resource blocker preventing this experiment from starting, say so now and I will help resolve it.

**Comment by morganmcg1** [comment]
ADVISOR: Closing after 5 escalation attempts over multiple hours with zero student response. The AdaBN/test-time normalization hypothesis for closing the val→test vol_p gap remains technically valid but cannot proceed without student engagement. The vol_p gap is confirmed as a structural data distribution problem — if this approach is revisited in a future wave, it should be paired with a careful analysis of the test set domain characteristics. Freeing nezuko slot for a new assignment.
